### ISSUE TITLE
Failing to compile TFLite MobileNetV3Small

### ISSUE THREAD

**Issue by GMNGeoffrey** [google/iree#7547]
I'm trying to port our MobileNetV3Small benchmark to compile directly from a tflite flatbuffer. Shockingly, there are issues! Currently hitting an error failing to legalize tosa.conv2d. The original TFLite flatbuffer is at https://tfhub.dev/google/lite-model/imagenet/mobilenet_v3_small_100_224/classification/5/default/1 (mirrored to https://storage.googleapis.com/iree-model-artifacts/mobilenet_v3_small_100_224.tflite). The mlir after iree-import-tflite is at https://gist.github.com/5a75b5cf8b2539c763c4bb2ba05cedf3

The error is `error: failed to legalize operation 'tosa.conv2d'` for this op:

```mlir
    %118 = "tosa.conv2d"(%117, %97, %96) {dilation = [1, 1], pad = [0, 1, 0, 1], stride = [2, 2]} : (tensor<?x224x224x3xf32>, tensor<16x3x3x3xf32>, tensor<16xf32>) -> tensor<?x112x112x16xf32>
```

**Comment by rsuderman** [comment]
Yeah, this is what i expected. This model has a dynamic batch dimension meaning we won't be able to use it for benchmarking if we pull from the tflite flatbuffer. There should be another tfhub model in iree-samples (https://github.com/google/iree-samples) that does not have the same issue. Though we may need to validate if the model is a proper mobilenet v3.

**Comment by GMNGeoffrey** [comment]
Yeah that's one of those mushroom identification models not published by Google. That would make it very different from the mobilenetv3 we're currently using. @antiagainst I think you were involved in selecting MobileNetV3Small as one of our benchmarks? Can you comment?

Also, why can't we handle a dynamic batch dimension? I keep hearing that dynamic shapes "just work", but then experiencing... not that

**Comment by rsuderman** [comment]
In this case the dynamic shape failure is on the TOSA end. When Tosa-To-Linalg was written we focused on statically shaped operations to simplify the compilation process. We are working on dynamic shapes more now however it is a non-trivial amount of work. I do not yet have a timeline for when we can support mobilenet models with dynamic dimensions.

**Comment by GMNGeoffrey** [comment]
Sorry if that sounded snippy :-) Totally ok that dynamic dims don't work in that layer. The thing I get grumpy about is when people *say* that they do, even when they don't

**Comment by antiagainst** [comment]
The MobileNetV3Small model is imported by calling Keras applications in the Python integration script. It actually specializes the batch and image dims to make them static:

https://github.com/google/iree/blob/64f9557e511f7d1ca7704789cb6a16c0aa414cf7/integrations/tensorflow/e2e/keras/applications/applications_test.py#L34-L35

This was way before when we weren't able to support dynamic shapes in the backends too; but right now supporting dynamic sizes should be fine (read: can compile and run) in the backends. (It might still hit bugs given I don't think we have tried dynamic shapes at a model level; we just have a few lit tests and small op runtime tests. Those bugs should be easy to fix though.) 

Right now actually to be performant, we cannot have all dimensions be dynamic. For those cases we'd need to have the full fledged tiling/padding/vectorization to expose static ops inside loop nests for vectorization. We aren't there yet. But having a single dynamic batch dimension should be fine; we should be able to generate performant kernels for it.

Back to swapping the models into TFLite sourced ones, I think we should use a "vanilla" MobileNetV3Small to replace the current one. I'm not clear how different the mushroom identification model is; but from the name I'd assume it differs a lot. So sounds it should be its own new series; it should not kill the existing one. Looks we still need to keep the current MobileNetV3Small from TF a bit longer..

**Comment by GMNGeoffrey** [comment]
> Back to swapping the models into TFLite sourced ones, I think we should use a "vanilla" MobileNetV3Small to replace the current one. I'm not clear how different the mushroom identification model is; but from the name I'd assume it differs a lot. 

Yeah that was my feeling as well.

> but right now supporting dynamic sizes should be fine (read: can compile and run) in the backends

Right, it sounds like the limitation here is in the TFLite->Tosa->linalg frontend, specifically. I think it will be fixed in the medium term, but we should find another options for now. I'd really like to get us completely off fetching artifacts from unstable formats and having to do out of band updates. Some options:

1. Save the TF model to SavedModel format and use that
2. Export the TF model to TFLite and use that
3. Manipulate the TFLite flatbuffer to make the batch size static.

The advantage to using a TFLite model is that then we can compare ourselves against TFLite, but if we use a SavedModel I think we can continue the current benchmark series as it really will be the same model.

> So sounds it should be its own new series; it should not kill the existing one.

I thought we agreed that this would be true of all the TFLite benchmarks replacing the TF ones?

**Comment by antiagainst** [comment]
> Right, it sounds like the limitation here is in the TFLite->Tosa->linalg frontend, specifically. I think it will be fixed in the medium term, but we should find another options for now. I'd really like to get us completely off fetching artifacts from unstable formats and having to do out of band updates. Some options:
> 
> 1. Save the TF model to SavedModel format and use that
> 2. Export the TF model to TFLite and use that
> 3. Manipulate the TFLite flatbuffer to make the batch size static.

Among these, my vote would be 2). I think we can already generate TFLite flatbuffers from the Python integration script. So it should be simple.

> I thought we agreed that this would be true of all the TFLite benchmarks replacing the TF ones?

Yup! :)

**Comment by GMNGeoffrey** [comment]
Oh option 4: Convert the model from https://tfhub.dev/google/imagenet/mobilenet_v3_small_100_224/classification/5 to TFLite with a static batch size of 1. Then we're using a model from tfhub instead of... whatever the hell the keras APIs are doing to fetch things from...somewhere? I'm guessing they're actually the same though. Not sure which would be more official. @phoenix-meadowlark any thoughts on whether we should use the tfhub one, the keras one, or whether they're the same?

> I think we can already generate TFLite flatbuffers from the Python integration script.

You mean the applications test running against tflite? Yeah I think I can grab it from the test artifacts

**Comment by GMNGeoffrey** [comment]
I've uploaded the artifact from running our tests at https://storage.googleapis.com/iree-model-artifacts/MobileNetV3SmallKeras.tflite

**Comment by phoenix-meadowlark** [comment]
They should represent the same computation, but how they represent it may be different (e.g. because of different python implementations or because of how the same python implementation was lowered). There's probably no way to cover all of that so I'd just go with whatever's easiest to maintain.

**Comment by GMNGeoffrey** [comment]
So I decided to take a peek at the json dump of each flatbuffer. AFAICT, the one from tfhub doesn't really have a dynamic batch dimension. It's only the function *signature* that does. Not super familiar with this setup, but seems like if it was passed anything other than 1 it would just crash?

```json
        {
          shape: [
            1,
            224,
            224,
            3
          ],
          buffer: 1,
          name: "serving_default_inputs:0",
          quantization: {
          },
          shape_signature: [
            -1,
            224,
            224,
            3
          ]
        },
```

There's a few places where there are dynamic signatures with static shapes (the file is too big to look at it all). Here's the full compressed [json dump](https://github.com/google/iree/files/7516999/mobilenet_v3_small_100_224.json.gz).

Not sure if this changes our thoughts

**Comment by antiagainst** [comment]
Not exactly sure how to modify TFLite flatbuffers.. I'd say just use the one generated from Keras for now. It's MobileNetV3Small. :)

**Comment by GMNGeoffrey** [comment]
Yeah I don't think we should manually modify the flatbuffer, but maybe we should export the tfhub TensorFlow model to tflite: https://tfhub.dev/google/imagenet/mobilenet_v3_small_100_224/classification/5. The question is whether we want to be able to link people to TFHub when we say what model we're using. Whereas pointing at https://github.com/google/iree/blob/main/integrations/tensorflow/e2e/keras/applications/applications_test.py is not quite as clean. Another option would be extracting whatever it is that initialize_model actually does in the specific case. I think something like `tf.keras.applications.MobileNetV3Small("imagenet")`. Basically I think it's valuable to provide a reproducible way that instructs someone how they could generate the same model

**Comment by GMNGeoffrey** [comment]
Ok I was able to reproduce the TFHub tflite model modulo some TF metadata and then nail down the input batch size with a bit of fiddling. I'm inclined to go with that, as it's a tflite model and we can show with a few lines of python how someone else could reproduce it. It also means that when the dynamic shape issues on the frontend are ironed out we could seamlessly switch to the dynamic batch version (literally the only substantive difference in the mlir is `?` vs `1`).

<details>
<summary>Details</summary>
Setup environment. Exact versions of things likely don't matter, but these match the ones in our docker images

```shell
python3.8 -m venv generate_mobilenetv3.venv
pip install --upgrade \
  tf-nightly==2.7.0.dev20210806 \
  keras-nightly==2.7.0.dev2021080600 \
  tensorflow-hub==0.12.0
```

```python
# generate_mobilenetv3small.py
import tensorflow as tf
from tf import keras
import tensorflow_hub as hub

m = tf.keras.Sequential([
    hub.KerasLayer("https://tfhub.dev/google/imagenet/mobilenet_v3_small_100_224/classification/5")
])
m.build([None, 224, 224, 3])
converter = tf.lite.TFLiteConverter.from_keras_model(m)
tflite_model = converter.convert()

# Save the model.
with open('MobileNetV3SmallDynamicBatch.tflite', 'wb') as f:
  f.write(tflite_model)

m = tf.keras.Sequential([
    keras.Input(shape=(224,224,3), batch_size=1),
    hub.KerasLayer("https://tfhub.dev/google/imagenet/mobilenet_v3_small_100_224/classification/5")
])
converter = tf.lite.TFLiteConverter.from_keras_model(m)
tflite_model = converter.convert()

# Save the model.
with open('MobileNetV3SmallStaticBatch.tflite', 'wb') as f:
  f.write(tflite_model)
```

```shell
python generate_mobilenetv3small.py
```

Let's check the differences between models are as expected. Going to use the MLIR since that format's way less verbose and more familiar to me than json flatbuffers

```shell
wget https://tfhub.dev/google/lite-model/imagenet/mobilenet_v3_small_100_224/classification/5/default/1?lite-format=tflite -O MobileNetV3SmallTFHub.tflite
./integrations/tensorflow/bazel-bin/iree_tf_compiler/iree-import-tflite MobileNetV3SmallStaticBatch.tflite -o MobileNetV3SmallStaticBatch.tflite.mlir
./integrations/tensorflow/bazel-bin/iree_tf_compiler/iree-import-tflite MobileNetV3SmallDynamicBatch.tflite -o MobileNetV3SmallDynamicBatch.tflite.mlir
./integrations/tensorflow/bazel-bin/iree_tf_compiler/iree-import-tflite MobileNetV3SmallTFHub.tflite -o MobileNetV3SmallTFHub.tflite.mlir
# Diff is just metadata
$ diff MobileNetV3SmallDynamicBatch.tflite.mlir MobileNetV3SmallTFHub.tflite.mlir 
2c2
<   func @main(%arg0: tensor<?x224x224x3xf32> {tf_saved_model.index_path = ["keras_layer_input"]}) -> (tensor<?x1001xf32> {tf_saved_model.index_path = ["keras_layer"]}) attributes {tf.entry_function = {inputs = "serving_default_keras_layer_input:0", outputs = "StatefulPartitionedCall:0"}, tf_saved_model.exported_names = ["serving_default"]} {
---
>   func @main(%arg0: tensor<?x224x224x3xf32> {tf_saved_model.index_path = ["inputs"]}) -> (tensor<?x1001xf32> {tf_saved_model.index_path = ["logits"]}) attributes {tf.entry_function = {inputs = "serving_default_inputs:0", outputs = "StatefulPartitionedCall:0"}, tf_saved_model.exported_names = ["serving_default"]} {

# Diff is metadata + ? -> 1
$ diff -U 1 <(cat MobileNetV3SmallDynamicBatch.tflite.mlir | tr '?' '1') MobileNetV3SmallStaticBatch.tflite.mlir 
--- /dev/fd/63  2021-11-11 16:19:42.023055300 -0800
+++ MobileNetV3SmallStaticBatch.tflite.mlir     2021-11-11 16:05:07.520007292 -0800
@@ -1,3 +1,3 @@
 module  {
-  func @main(%arg0: tensor<1x224x224x3xf32> {tf_saved_model.index_path = ["keras_layer_input"]}) -> (tensor<1x1001xf32> {tf_saved_model.index_path = ["keras_layer"]}) attributes {tf.entry_function = {inputs = "serving_default_keras_layer_input:0", outputs = "StatefulPartitionedCall:0"}, tf_saved_model.exported_names = ["serving_default"]} {
+  func @main(%arg0: tensor<1x224x224x3xf32> {tf_saved_model.index_path = ["input_1"]}) -> (tensor<1x1001xf32> {tf_saved_model.index_path = ["keras_layer_1"]}) attributes {tf.entry_function = {inputs = "serving_default_input_1:0", outputs = "StatefulPartitionedCall:0"}, tf_saved_model.exported_names = ["serving_default"]} {
     %0 = "tosa.const"() {value = dense<6.000000e+00> : tensor<f32>} : () -> tensor<f32>
@@ -364,3 +364,3 @@
     %361 = "tosa.conv2d"(%360, %2, %1) {dilation = [1, 1], pad = [0, 0, 0, 0], stride = [1, 1]} : (tensor<1x1x1x1024xf32>, tensor<1001x1x1x1024xf32>, tensor<1001xf32>) -> tensor<1x1x1x1001xf32>
-    %362 = "tosa.reshape"(%361) {new_shape = [-1, 1001]} : (tensor<1x1x1x1001xf32>) -> tensor<1x1001xf32>
+    %362 = "tosa.reshape"(%361) {new_shape = [1, 1001]} : (tensor<1x1x1x1001xf32>) -> tensor<1x1001xf32>
     return %362 : tensor<1x1001xf32>
```

I'm pretty satisfied with that
</details>

**Comment by antiagainst** [comment]
👍 👍 👍  Ship it! :D
